internal/resource/url.go

```diff
@@ -375,6 +388,36 @@ type s3target interface {
 	io.ReadSeeker
 }
 
+// getBucketRegion determines the partition and region a given S3 bucket is located in
+func (f *Fetcher) getBucketRegion(ctx context.Context, sess client.ConfigProvider, u url.URL) (string, error) {
+	// GetBucketRegion will return NotFound if the bucket doesn't exist in the same
+	// partition as the regionHint despite the default resolver in the session object
+	// supporting all partitions. First attempt to fetch from the S3RegionHint value
+	// on the Fetcher object if it exists, if not fallthrough to try with one region
+	// from each partition.
+	if f.S3RegionHint != "" {
+		region, err := s3manager.GetBucketRegion(ctx, sess, u.Host, f.S3RegionHint)
+		if err != nil {
+			if aerr, ok := err.(awserr.Error); ok && aerr.Code() != "NotFound" {
+				return "", err
+			}
+		} else {
+			return region, nil
+		}
+	}
+	for _, regionHint := range []string{"us-east-1", "us-gov-east-1", "cn-north-1"} {
```

Comment:
Do we expect a case where the region of a bucket exists in a place that's tied to a region hint of say `xyz` which isn't present in the `regionHint` array?

Reply:
I'm not sure I understand your question; does this clarify it?

This list contains one region from each partition. The reason that we're doing this loop is that the [`s3manager.GetBucketRegion` call will fail with a `NotFound` error if the `regionHint` is not in the same partition as the bucket](https://docs.aws.amazon.com/sdk-for-go/api/service/s3/s3manager/#GetBucketRegion), so the new ordering we're trying is:

1. If being ran on AWS we first attempt with the region the server is located in
1. `us-east-1` (US partition)
1. `us-gov-east-1` (Govcloud)
1. `cn-north-1` (aws-cn)

Reply:
> I'm not sure I understand your question; does this clarify it?

That question was more on this information `For example to get the region of a bucket which exists in "eu-central-1" you could provide a region hint of "us-west-2".`